File: Mobile/lib/service/report_service.dart
import 'dart:convert';

import 'package:gowheel_flutterflow_ui/service/storage_service.dart';
import 'package:gowheel_flutterflow_ui/url.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

import '../models/report_model.dart';

class ReportService{
  static final ReportService _instance = ReportService._internal();
  factory ReportService() => _instance;
  ReportService._internal();

  final TokenService tokenService = TokenService();

  Future<List<Report>> getAllReportType() async {
    try{
      final response = await http.get(
        Uri.parse("${URL.baseUrl}Admin/ReportType/GetAll"),
        headers: {
          'Content-Type': 'application/json',
        },
      );
      if (response.statusCode == 200) {
        final Map<String, dynamic> responseData = json.decode(response.body);

        if (responseData['success'] != true || responseData['data'] == null) {
          return [];
        }

        final List<dynamic> postsJson = responseData['data'] as List<dynamic>;
        final posts = postsJson.map((json) {
          try {
            return Report.fromJson(json);
          } catch (e) {
            return null;
          }
        }).whereType<Report>().toList();

        return posts;
      } else {
        throw Exception('Failed to load report type');
      }
    } catch (e) {
      throw Exception('Failed to load report type: $e');
    }
  }
  Future<bool> createReport({
    required String content,
    required int postId,
    required int reportTypeId,
  }) async {
    try {
      final token = await getToken();

      final response = await http.post(
        Uri.parse("${URL.baseUrl}UserReport/ReportByPostId"),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: json.encode({
          'id': 0,
          'content': content,
          'postId': postId,
          'reportTypeId': reportTypeId,
        }),
      );

      if (response.statusCode == 200) {
        final Map<String, dynamic> responseData = json.decode(response.body);
        return responseData['success'] == true;
      } else {
        return false;
      }
    } catch (e) {
      return false;
    }
  }

  Future<String?> getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('jwt_token');
  }
}

File: FE/src/Service/api/ReportService.js
import axios from "axios";
import ReportDTO from "../../DTOs/ReportDto";
const ReportService = {
  async getAllReport() {
    const token = sessionStorage.getItem("authToken");
    const response = await axios.get(
      "http://localhost:5027/api/ManagerReport/GetAll",
      {
        headers: {
          Authorization: `Bearer ${token}`,
          "Content-Type": "application/json",
        },
      }
    );
    return response.data;
  },
  async getReportById(id) {
    const token = sessionStorage.getItem("authToken");
    const response = await axios.get(
      `http://localhost:5027/api/ManagerReport/GetById/${id}`
    );
    console.log(response);
    if (response.status !== 200) {
      throw new Error(`Error: ${response.status} - ${response.statusText}`);
    }

    console.log("Report by id: ", response);
    return response.data;
  },
  async AddReport(reportDto) {
    const token = sessionStorage.getItem("authToken");
    const formData = new ReportDTO(reportDto).toFormData();
    const response = await axios.post(
      "http://localhost:5027/api/UserReport/ReportByPostId",
      formData,
      {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      }
    );
    console.log(response);
    if (response.status !== 200) {
      console.log("Error: ", response);
    }

    return response.data;
  },

  async SubmitReport(id, bien) {
    console.log(id,bien);
      const formData = new FormData();
      formData.append("id", id);
      formData.append("isAccept", bien);

      // Lấy token từ sessionStorage
      const token = sessionStorage.getItem("authToken");
      const response = await fetch(
        `http://localhost:5027/api/ManagerReport/ExamineReport/${id}`,
        {
          method: "PUT",
          headers: {
            Authorization: `Bearer ${token}`
          },
          body: formData,
        }
      );
      console.log(response);
      return response;
  },

  
};
export default ReportService;
